File: core/src/main/java/org/aresclient/ares/api/command/Command.java
package org.aresclient.ares.api.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.CommandNode;
import net.meshmc.mesh.loader.Mod;

import java.util.Collection;
import java.util.Collections;

public class Command {
    public interface IContext {
        void print(String message);
        void error(String message);
    }

    @Mod.Interface
    private static Command.IContext CONTEXT;

    private static final CommandDispatcher<IContext> DISPATCHER = new CommandDispatcher<>();

    private final CommandNode<IContext> node;

    public Command(CommandNode<IContext> node) {
        this.node = node;
    }

    public CommandNode<IContext> getNode() {
        return node;
    }

    public Collection<String> getUsages(IContext context) {
        return getUsages(context, node);
    }

    public static void execute(String command) {
        try {
            DISPATCHER.execute(command, CONTEXT);
        } catch(CommandSyntaxException e) {
            CONTEXT.error(e.getLocalizedMessage());
        }
    }

    public static CommandNode<IContext> register(LiteralArgumentBuilder<IContext> builder) {
        return DISPATCHER.register(builder);
    }

    public static Collection<String> getUsages(IContext context, CommandNode<IContext> node) {
        return DISPATCHER.getSmartUsage(node, context).values();
    }

    public static CommandNode<IContext> getCommand(Collection<String> path) {
        return DISPATCHER.findNode(path);
    }

    public static CommandNode<IContext> getCommand(String name) {
        return getCommand(Collections.singletonList(name));
    }
}

File: src/main/kotlin/org/aresclient/ares/api/command/Command.kt
package org.aresclient.ares.api.command

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.tree.CommandNode

open class Command(private val node:CommandNode<IContext?>?) {
	interface IContext {
		fun print(message:String?)
		fun error(message:String?)
	}

	companion object {
		@JvmStatic private val CONTEXT:IContext? = null
		@JvmStatic private val DISPATCHER:CommandDispatcher<IContext?> = CommandDispatcher()

		fun register(builder:LiteralArgumentBuilder<IContext?>?): CommandNode<IContext?> {
			return DISPATCHER.register(builder)
		}

		fun getUsages(context:IContext?, node:CommandNode<IContext?>?): Collection<String> {
			return DISPATCHER.getSmartUsage(node, context).values
		}

		fun getCommand(path:Collection<String?>?):CommandNode<IContext?> {
			return DISPATCHER.findNode(path)
		}

		fun getCommand(name:String):CommandNode<IContext?> {
			return getCommand(listOf(name))
		}
	}

	fun getNode():CommandNode<IContext?>? {
		return node
	}

	fun getUsages(context:IContext?):Collection<String> {
		return getUsages(context, node)
	}

	fun execute(command:String?) {
		try {
			DISPATCHER.execute(command, CONTEXT)
		} catch (e:CommandSyntaxException) {
			CONTEXT!!.error(e.localizedMessage)
		}
	}
}

